tests/k4unit.q

```diff
@@ -61,50 +62,82 @@ KUltf:{ / (load test file) - load tests in csv file <x> into KUT
 	before:count KUT;
-	files:(` sv)each(x,'key x);KUltf each files where(lower files)like"*.csv";
+	files:(` sv)each(x,'key x);KUltf each files where 1=sum (lower files) like/: ("*process.csv";"*.csv");
 	neg before-count KUT}
 
 KUrt:{ / (run tests) - run contents of KUT, save results to KUTR
+	update csvline:(raze value exec 2+til count i by file from KUT) from `KUT;
 	before:count KUTR;uf:exec asc distinct file from KUT;i:0;
 	if[.KU.VERBOSE;.lg.o[`k4unit;"start"]];
-	exec KUexec'[lang;code;repeat] from KUT where action=`beforeany;
+	KUerrparse[`beforeany;] exec KUexec'[lang;code;repeat],file,csvline from KUT where action=`beforeany;
 	do[count uf;
 		ufi:uf[i];KUTI:select from KUT where file=ufi;
 		if[.KU.VERBOSE;.lg.o[`k4unit;(string ufi)," ",(string exec count i from KUTI where action in `run`true`fail)," test(s)"]];
-		exec KUexec'[lang;code;repeat] from KUT where action=`beforeeach;
-		exec KUexec'[lang;code;repeat] from KUTI where action=`before;
+		KUerrparse[`beforeach;] exec KUexec'[lang;code;repeat],file,csvline from KUT where action=`beforeeach;
+		KUerrparse[`before;] exec KUexec'[lang;code;repeat],file,csvline from KUTI where action=`before;
 		/ preserve run,true,fail order
-		exec KUact'[action;lang;code;repeat;ms;bytes;file]from KUTI where action in`true`fail`run;
-		exec KUexec'[lang;code;repeat] from KUTI where action=`after;
-		exec KUexec'[lang;code;repeat] from KUT where action=`aftereach;
+		exec KUact'[action;lang;code;repeat;ms;bytes;file;csvline] from KUTI where action in`true`fail`run;
+		KUerrparse[`after;] exec KUexec'[lang;code;repeat],file,csvline from KUTI where action=`after;
+		KUerrparse[`aftereach;] exec KUexec'[lang;code;repeat],file,csvline from KUT where action=`aftereach;
 		i+:1];
-	exec KUexec'[lang;code;repeat] from KUT where action=`afterall;
+	KUerrparse[`afterall;] exec KUexec'[lang;code;repeat],file,csvline from KUT where action=`afterall;
 	if[.KU.VERBOSE;.lg.o[`k4unit;"end"]];
 	neg before-count KUTR}
 
 KUpexec:{[prefix;lang;code;repeat;allowfail]
 	s:(string lang),")",prefix,$[1=repeat;string code;"do[",(string repeat),";",(string code),"]"];
-	if[1<.KU.VERBOSE;.lg.o[`k4unit;s]];$[.KU.DEBUG&allowfail;value s;@[value;s;`FA1L]]}
-
-KUfailed:{`FA1L~x}
+	if[1<.KU.VERBOSE;.lg.o[`k4unit;s]];
+	$[.KU.DEBUG & allowfail;value s;@[value;s;{(`err;`$x;y)}[;code]]]
+	}
 
+// If in error - it now returns the error as well as the offending code - basically the same function as KUpexec?
 KUexec:{[lang;code;repeat]
-	value(string lang),")",$[1=repeat;string code;"do[",(string repeat),";",(string code),"]"]}
+	strexec:(string lang),")",$[1=repeat;string code;"do[",(string repeat),";",(string code),"]"];
+	@[value;strexec;{(`err;`$x;y)}[;code]]
+	}
+
+// Generate error logs from beforeeach, before, after and aftereach tests
+KUerrparse:{[action;out]
+	vals:1_' out where `err ~/: first each out:raze each flip value out;
+	.lg.e[`KUexecerr;] each KUerrparseinner[action;;;;] .' vals;
```

Comment:
given that k4unit.q is automatically loaded & tests run as torq.q loads (when `-test` arg is present), I think `.lg.e` would cause torq to exit if not run in debug mode? Is that the expected behaviour?

Reply:
When not run in debug mode any errors in the test code itself is logged to an err_proc.log file. The .lg.e seems to work fine

Reply:
Ah yep, looking at it now I see `.proc.initialised:1b` is set before loading & running tests, so `.lg.e` won't exit :+1: